code/missionui/missiondebrief.cpp

```diff
@@ -952,10 +952,11 @@ void debrief_choose_voice(char *voice_dest, size_t buf_size, char *voice_base, i
 void debrief_choose_medal_variant(char *buf, int medal_earned, int zero_based_version_index)
 {
 	Assert(buf != NULL && medal_earned >= 0 && zero_based_version_index >= 0);
+	Assertion(strlen(Resolution_prefixes[gr_screen.res]) + strlen(Medals[medal_earned].debrief_bitmap) < MAX_FILENAME_LEN, "The filename for %s is over the filename length limit of 31 characters.  On release builds, this filename will be shortened and will not work.", buf);
 
 	// start with the regular file name, adapted for resolution
-	sprintf(buf, NOX("%s%s"), Resolution_prefixes[gr_screen.res], Medals[medal_earned].debrief_bitmap);
-
+	snprintf(buf, MAX_FILENAME_LEN, NOX("%s%s"), Resolution_prefixes[gr_screen.res], Medals[medal_earned].debrief_bitmap);
```

Comment:
I think it should be `MAX_FILENAME_LEN - 1`, judging from the other uses of snprintf.  `MAX_FILENAME_LEN` is the buffer size, not the number of non-null characters.  (There are several naming inaccuracies like this throughout the code.)

Reply:
Ok, I checked documentation and wrote a test program.  It looks like the size argument should be the length of the destination string and should not adjust for the null terminator.  snprintf will place a null terminator at the end, even if the source string is longer than the destination string.  The only problematic thing that can be done is to specify a length longer than the destination string.  Then the destination string will be null terminated at the start.


![Screen Shot 2023-11-12 at 10 52 44 PM](https://github.com/scp-fs2open/fs2open.github.com/assets/12529946/909bc4d7-febd-489f-a9a4-80b255abd974)

Reply:
Well it does look like `snprintf` treats its *n* argument differently than `strncpy` does...
> snprintf() copies at most n – 1 characters into the output buffer and then appends a terminating null character.

and
> Writes the results to a character string buffer. At most bufsz - 1 characters are written. The resulting character string will be terminated with a null character.

This means that `MAX_FILENAME_LEN` is, in fact, the correct number to use, and therefore quite a few instances of `snprintf` in the codebase shortchange their strings by 1 char.